Add message reaction functionality

Neriderc opened the pull request:
This pull request adds the ability to "react" to messages (posts and replies), using this emoji picker: https://github.com/ealush/emoji-picker-react

Because of the size of the library (apparently adds 250kb to front end) and my understanding of wanting to keep the frontend light, this is DISABLED BY DEFAULT. You need to set `ALLOW_MESSAGE_REACTIONS="true"` in the .env

Reactions are shown under a message/above the footer. Click "+ React" button to show emoji picker. Select an emoji to react to the post with it.

Emojis are summarised by type (e.g. " :slightly_smiling_face: 3") and if the current user has selected this emoji then they can click on it to undo their reaction. Hovering over an emoji in the list of reactions will show the users that reacted with that reaction. Selecting a reaction in the emoji picker that you have already used will undo the reaction.

One user can select multiple emoji of different types but not the same one twice.

This creates a new entity Reaction and so requires a database migration. It also adds new frontend components, new API endpoints, etc.

Resolves #83

As always, happy to hear suggestions!

nrobinaubertin commented on api/src/Controller/Info.php:
```diff
@@ -46,6 +46,7 @@ public function index(): Response
           'default_lang' => $this->getParameter('lang'),
           'allow_email' => 'true' == $this->getParameter('allow.email'),
           'allow_public_links' => 'true' == $this->getParameter('allow.public.links'),
+          'allow_message_reactions' => 'true' == $this->getParameter('allow.message.reactions'),
```
I like the possibility of disabling reactions but you've done it only on the backend side. `emoji-picker-react` is still going to be compiled as part of the webapp and pushed to the user.  
Ideally, the frontend would check if this parameter is enabled and only load the emoji picker code if it is but doing so is more complicated.

Neriderc replied:
The `import EmojiPicker from "emoji-picker-react"' is within a separate component. My understanding was that because this component is not loaded because of the `&& api?.info?.allow_message_reactions` check, this would mean `emoji-picker-react` wouldn't be pushed to the user?

I'm new to this so happy to take advice on making this better.

nrobinaubertin replied:
It will not display it but the code is still here (and has to be downloaded and parsed by the client). `import` is resolved during transpilation (to create a javascript payload that will be send to the browser) and the `&& api?.info?.allow_message_reactions` is at runtime.  

That said, it's possible that parcel is doing some magic that I'm not aware of. I'll check exactly what's happening when trying your PR locally

Neriderc replied:
You were right, it's still bundled as part of the overall js package (to the tune of 250kb or so). I think I've got this sorted now though.

I've pushed a new commit that uses lazy loading. I came across a [bug in preact](https://github.com/preactjs/preact/issues/2519) so I have run `npm update` to update packages to latest minor versions, which has resolved the bug. The commit includes the updated package.json/lock files.

Through some trial and error I've worked out that importing on mount of the component (instead of waiting for first use) means the user doesn't have to mess around clicking the button a few times when they first try to select a reaction.

Again - I'm no expert, so please let me know if something needs changing.

nrobinaubertin replied:
From my testing, it works ! Thanks a lot for that

nrobinaubertin reviewed (commented):
A first review: this seems like a nice addition and I like the way you've implemented it.  
I still want to test it locally and I need to review the frontend code.

Neriderc commented on api/src/Entity/Reaction.php:
```diff
@@ -0,0 +1,134 @@
+     *
+     * @ORM\Column(type="string")
+     */
+    private $reaction;
```
Fair point, $value sounds good. I'll update it.

Neriderc replied:
I've just pushed a commit that makes this change.

Neriderc commented on api/src/Controller/Message/Delete.php:
```diff
@@ -59,6 +60,11 @@ public function index(
             $this->em->remove($n);
         }
 
+        $reactions = $message->getReactions();
+        foreach ($reactions as $reaction) {
+            $this->em->remove($reaction);
+        }
```
I tried and gave up and copied what you had for notifications. I couldn't work out how to get doctrine to apply the cascade. I can try again but happy for advice!

nrobinaubertin replied:
I'll try to make it work once I'll test it locally

Neriderc replied:
Not sure what the issue was the first time I tried but I have got cascade delete working now in a new commit.

Neriderc commented:
I haven't yet generated a new migration but I can. Do we typically want many migrations, one for each change (e.g. one for this reaction function, one for email refactor, etc), or do we want to do one big migration that goes from the previous version to the new version?

nrobinaubertin commented:
> I haven't yet generated a new migration but I can. Do we typically want many migrations, one for each change (e.g. one for this reaction function, one for email refactor, etc), or do we want to do one big migration that goes from the previous version to the new version?

One migration for each PR is the best IMO. It is easier for us to handle

nrobinaubertin commented:
Thank you very much for this PR @Neriderc. I like it a lot.  
So much in fact that I wanted to fiddle with it right away. I've re-styled a bit and moved the "react" button to the footer of each message in the following branch: https://github.com/zusam/zusam/tree/post-reactions-footer

WDYT ? If you think that it's a good change, we can add this to the PR

Neriderc commented:
I haven't tried it just yet but I had it in the footer originally, I moved it into the message body because I thought the replies under a post looked off as these only have the message body white and the footer/etc is a grey colour. Have you tested with post replies to see how it looks?

nrobinaubertin commented:
> I haven't tried it just yet but I had it in the footer originally, I moved it into the message body because I thought the replies under a post looked off as these only have the message body white and the footer/etc is a grey colour. Have you tested with post replies to see how it looks?

It's maybe a bit tiny but I think that it looks good otherwise
![image](https://github.com/user-attachments/assets/1dd40a6d-9602-4aa7-a995-80de127b899e)

Neriderc commented:
I've had a chance to try yours out. I think it looks good, works well. Looks a bit funny on a mobile screen though, the edit function is collapsed to the menu but it seems to be missing a dot?

![image](https://github.com/user-attachments/assets/0669ebf8-efd5-405b-b0df-928c13dde47c)

Neriderc commented:
I've added your commit into this branch. It looks like the issue I posted above is where you've copied the bullet from the edit section, and that has classes causing it to hide when the screen gets small. I've updated the bullet classes to be consistent with the other ones, and it seems to work properly now.

I see you've also updated the migration.

Is there anything else needed for this pull request?

nrobinaubertin commented on app/src/message/message-reactions.component.js:
```diff
@@ -0,0 +1,77 @@
+          style={{
+            cursor: currentUserReactionId ? "pointer" : "default",
+          }}
+          title={users.slice(0, MAX_VISIBLE_USERS).join("\n") + (users.length > MAX_VISIBLE_USERS ? `\nand ${users.length - MAX_VISIBLE_USERS} more...` : "")}
```
I saw that while editing on my branch but thought to leave for later (so now): we would need to add a translation string for this. It could have only a key for english for now though.

Neriderc replied:
Good catch! I've just added the en_US translation and the call to the translation function in a new commit.

nrobinaubertin reviewed (commented):
Last comment and we're good ! (also a linting pass for the CI)

nrobinaubertin commented on app/src/message/message-reactions.component.js:
```diff
@@ -47,7 +49,7 @@ export default function MessageReactions(props) {
             cursor: currentUserReactionId ? "pointer" : "default",
           }}
-          title={users.slice(0, MAX_VISIBLE_USERS).join("\n") + (users.length > MAX_VISIBLE_USERS ? `\nand ${users.length - MAX_VISIBLE_USERS} more...` : "")}
+          title={users.slice(0, MAX_VISIBLE_USERS).join("\n") + (users.length > MAX_VISIBLE_USERS ? `\nand ${users.length - MAX_VISIBLE_USERS} ${t("more")}...` : "")}
```
The `and` is missing from the translation. I think that you should use [interpolation](https://www.i18next.com/translation-function/interpolation#basic) here

Neriderc replied:
Ah sorry, dumb mistake. I completely missed that the "and" was there. Should be fixed now.

nrobinaubertin commented:
Just some linting errors that can be fixed with `php bin/composer fix` and this is ready to be merged

nrobinaubertin commented:
Thank you for the reactions support !

Neriderc commented:
Thanks for merging!
